README.md
# Robson Analyzer
An extension deisgned around the will to make you suffer but no just suffer, suffer with Robson

## Features

#### This extension provides a lot o features like: 
> - Autocomplete for you to do things fast and suffer later
> - Syntax highlighting to trick your brain to thinking that what your building is worth it
> - Error catching to make you a false sense o security
> - Inlay hints to make your already noisy workspace from robson's noisier

## Requirements

Required to have enough time to waste 

## Known Issues

This extension is perfect



CHANGELOG.md
# Change Log

## 0.1.2
> #### Added
> - path autocomplete
> - syntax highligh on robsons

> #### Fixed
> - inlay hints appearing where they shouldn't

## 0.1.1
> #### Changed
> - now inlay hints show variations of the same command
> - .robson icon

## 0.1.0
> #### Added
> - unknown inlay hint
> - support for the new commands in robson 0.1.1

> #### Changed
> - build task to actually build

## 0.0.9
> #### Changed
> - run task to support windows computers
> - README.md to update the view of the new features

## 0.0.8
> #### Added
> - error for comeu, fudeu, penetrou and chupou malformated numbers
> - warning for unused aliases
> - add build task to run the code (one wich doesnt build anything)

## 0.0.7

> #### Added
> - error for unknown keywords
> - error for missing parameters

## 0.0.6

> #### Added
> - inlay hints
> - go to definition on aliases

## 0.0.5

> #### Added
> - alias autocomplete
> - keyword autocomplete 
> - lambeu malformated value error
> - camel case warning and autocomplete

## 0.0.4

> #### Added
> - icon for the extension
> - support for the penetrou keyword

## Initial release

#### Added syntax highlights

> #### Added snippets:
> - add
> - sub
> - push
> - if
> - vstack
> - input
> - print
> - pnumber
> - jump
> - set
> - pop
> - load

#### Added icon for .robson



server/src/utils.ts
export function lowerAlias(to_lower: string): string {
    let lowercase = to_lower.replace(/([A-Z])/g, "_$1").toLocaleLowerCase();
    if (lowercase[0] === "_") {
        lowercase = lowercase.slice(1, lowercase.length);
    }
    if (lowercase[0] === ":") {
        let sliced = lowercase.slice(1, lowercase.length);
        if (sliced[0] == "_") {
            lowercase = `:${sliced.slice(1, sliced.length)}`
        }
    }
    return lowercase
}
export function removeComments(string: string): string {
    const removed_comments = string.split(";");
    if (removed_comments.length > 0) {
        string = removed_comments[0];
    }
    return string.trim()
}

export function remove_file_prefix(a: string): string {
    let splited = a.split("file://");
    if (splited.length < 2) {
        return a;
    }
    let new_path = "";
    for (let i = 1; i < splited.length; i++) {
        new_path += splited[i];
    }
    return new_path;
}

export function get_robsons_path(a: string): string[] {
    const splited = a.split("robsons");

    if (splited.length != 2) {
        return [];
    }

    return splited[1].split("\\");

}

export function getOpcode(string: string): number {
    let opcode = 0;
    const removed_comments = removeComments(string);
    const splited = string.trim().split(' ');
    for (const keyword of splited) {
        if (keyword.trim() == "robson") {
            opcode++;
        } else {
            opcode = -1;
            break;
        }
    }
    return opcode;
}
function validParam() {

}
export const parameterCount = {
    1: 3,
    2: 3,
    3: 1,
    4: 3,
    5: 1,
    6: 3,
    7: 0,
    8: 0,
    9: 1,
    10: 1,
    11: 0,
    12: 1,
    13: 1,
    14: 0,
    15: 1
}
export const labels = {
    1: "operation",
    2: "if lower",
    3: "push",
    4: "if equal",
    5: "vstack",
    6: "input",
    7: "print",
    8: "pnumber",
    9: "jump",
    10: "set",
    11: "pop",
    12: "load"
}

server/src/server.ts
import {
    createConnection,
    TextDocuments,
    ProposedFeatures,
    InitializeParams,
    DidChangeConfigurationNotification,
    CompletionItem,
    CompletionItemKind,
    TextDocumentPositionParams,
    TextDocumentSyncKind,
    InitializeResult,
    CodeAction,
    CodeActionKind,
    Command,
    TextDocumentEdit,
    TextEdit,
    Location,
    Range,
    Hover,
} from 'vscode-languageserver/node';
import { TextDocument } from 'vscode-languageserver-textdocument';
import { get_robsons_path, lowerAlias, removeComments, remove_file_prefix } from "./utils";
import { validateTextDocument } from './validations';
import { ConnectionOptions } from 'vscode-languageclient';
import * as fs from "node:fs/promises";
import * as path from "path"
interface DefaultSettings {
    maxNumberOfProblems: number;
}

export type AliasInfo = {
    range: Range,
    value: string
}

const defaultSettings: DefaultSettings = { maxNumberOfProblems: 1000 };

export let connection = createConnection(ProposedFeatures.all);
export let hasDiagnosticRelatedInformationCapability: boolean = false;

let documents: TextDocuments<TextDocument> = new TextDocuments(TextDocument);
let aliases: AliasInfo[] = [];
let aliasUse: AliasInfo[] = [];
let globalSettings: DefaultSettings = defaultSettings;

let documentSettings: Map<string, Thenable<DefaultSettings>> = new Map();
let hasConfigurationCapability: boolean = false;

export function setAliases(newAliases: AliasInfo[]) {
    aliases = newAliases;
}
export function setAliasUse(newUses: AliasInfo[]) {
    aliasUse = newUses;
}

connection.onInitialize((params: InitializeParams) => {
    let capabilities = params.capabilities;
    // Does the client support the `workspace/configuration` request?
    // If not, we fall back using global settings.
    hasConfigurationCapability = !!(
        capabilities.workspace && !!capabilities.workspace.configuration
    );

    hasDiagnosticRelatedInformationCapability = !!(
        capabilities.textDocument &&
        capabilities.textDocument.publishDiagnostics &&
        capabilities.textDocument.publishDiagnostics.relatedInformation
    );

    const result: InitializeResult = {
        capabilities: {
            textDocumentSync: TextDocumentSyncKind.Incremental,
            // Tell the client that this server supports code completion.
            completionProvider: {
                resolveProvider: true,
            },
            // Tell the client that this server support code actions.
            codeActionProvider: true,
            // Tell the client that this server support commnds
            executeCommandProvider: {
                commands: ["re.name"]
            },
            // Tell the client thtat this server support definitions
            definitionProvider: true,
        }
    };
    return result;
});

connection.onInitialized((t) => {
    if (hasConfigurationCapability) {
        // Register for all configuration changes.
        connection.client.register(DidChangeConfigurationNotification.type, undefined);
    }
});


export function getDocumentSettings(resource: string): Thenable<DefaultSettings> {
    if (!hasConfigurationCapability) {
        return Promise.resolve(globalSettings);
    }
    let result = documentSettings.get(resource);
    if (!result) {
        result = connection.workspace.getConfiguration({
            scopeUri: resource,
            section: 'RobsonAnalyzer'
        });
        documentSettings.set(resource, result);
    }
    return result;
}



connection.onDidChangeConfiguration(change => {
    if (hasConfigurationCapability) {
        // Reset all cached document settings
        documentSettings.clear();
    } else {
        globalSettings = <DefaultSettings>(
            (change.settings.languageServerExample || defaultSettings)
        );
    }

    // Revalidate all open text documents
    documents.all().forEach(validateTextDocument);
});

// Only keep settings for open documents
documents.onDidClose(e => {
    documentSettings.delete(e.document.uri);
});

// The content of a text document has changed. This event is emitted
// when the text document first opened or when its content has changed.
documents.onDidChangeContent(change => {
    validateTextDocument(change.document);
});

connection.onDidChangeWatchedFiles(_change => {
    // Monitored files have change in VS Code
    connection.console.log('We received a file change event');
});

connection.onDefinition(params => {
    const textDocument = documents.get(params.textDocument.uri);
    if (!textDocument) {
        return
    }
    let aliasSelected: AliasInfo | undefined;

    for (const alias of aliasUse) {
        const line = alias.range.start.line;
        const char_start = alias.range.start.character;
        const char_end = alias.range.end.character;
        if (line === params.position.line && params.position.character >= char_start && params.position.character <= char_end) {
            aliasSelected = alias;
        }
    }

    if (!aliasSelected) {
        return
    }
    const aliasDeclaration = aliases.find(a => a.value === aliasSelected?.value);
    if (!aliasDeclaration) {
        return
    }

    return Location.create(params.textDocument.uri, aliasDeclaration.range);
})
// connection.workspace.onDidChangeWorkspaceFolders((param) => {
//     // connection.console.log(param.added);

//     for (const a of param.added) {
//         connection.console.log(a.name);
//     }

// })
// This handler provides the initial list of the completion items.

connection.onCompletion(
    async (textDocumentPosition: TextDocumentPositionParams): Promise<CompletionItem[]> => {
        // The pass parameter contains the position of the text document in
        // which code complete got requested. For the example we ignore this
        // info and always provide the same completion items.
        // connection.console.log(textDocumentPosition.position.line)

        const textDocument = documents.get(textDocumentPosition.textDocument.uri);

        const raw_text = textDocument?.getText({
            start: {
                character: 0,
                line: textDocumentPosition.position.line
            },
            end: {
                character: 0,
                line: textDocumentPosition.position.line + 1
            }
        })
        const paths = [];
        if (raw_text) {
            const text = removeComments(raw_text);
            if (text.includes("robsons")) {
                const robsons_path = get_robsons_path(text);
                connection.console.log(`robson -> ${robsons_path}`);
                const folderConnection = await connection.workspace.getWorkspaceFolders();

                if (folderConnection) {
                    for (const folder of folderConnection) {
                        let root_path = remove_file_prefix(folder.uri);

                        robsons_path.forEach((x, i) => {
                            if (robsons_path.length - 1 !== i) {
                                root_path = path.join(root_path, x.trim());
                            }
                        });
                        try {
                            connection.console.log(`${robsons_path},, ${root_path}`);
                            const j = await fs.readdir(root_path);
                            for (const s of j) {
                                const info = await fs.lstat(path.join(root_path, s));
                                if (info.isFile()) {
                                    paths.push(s);
                                }
                                if (info.isDirectory()) {
                                    paths.push(`${s}\\`);
                                }
                                // connection.console.log(s);
                            }
                        } catch { }
                    }
                }
            }
        }
        const pathsCompletion: CompletionItem[] = paths.map(path => {
            return {
                label: path,
                kind: CompletionItemKind.File
            }
        })

        const aliasesCompletetion: CompletionItem[] = aliases.map(alias => {
            return {
                label: alias.value,
                kind: CompletionItemKind.Function
            }
        });
        if (pathsCompletion.length > 0) {
            return pathsCompletion
        }
        return [
            {
                label: 'comeu',
                kind: CompletionItemKind.TypeParameter,
                data: "b1"
            },
            {
                label: 'fudeu',
                kind: CompletionItemKind.TypeParameter,
                data: "b2"
            },
            {
                label: 'penetrou',
                kind: CompletionItemKind.TypeParameter,
                data: "b3"
            },
            {
                label: 'chupou',
                kind: CompletionItemKind.TypeParameter,
                data: "b4"
            },
            {
                label: 'lambeu',
                kind: CompletionItemKind.TypeParameter,
                data: "b5"
            },
            {
                label: 'robson',
                kind: CompletionItemKind.Keyword,
                data: "b6",
            },
            ...aliasesCompletetion,
        ];
    }
);

// This handler resolves additional information for the item selected in
// the completion list.
connection.onCompletionResolve(
    (item: CompletionItem): CompletionItem => {
        const datas: any = {
            "b1": () => {
                item.detail = 'comeu';
                item.documentation = 'returns the raw number';
            },
            "b2": () => {
                item.detail = 'fudeu';
                item.documentation = 'return the value from the given address';
            },
            "b3": () => {
                item.detail = 'penetrou';
                item.documentation = 'return the value of the adress that the given address points to';
            },
            "b4": () => {
                item.detail = 'chupou';
                item.documentation = 'return the value inside of the by the given amount'
            },
            "b5": () => {
                item.detail = "lambeu";
                item.documentation = "return the line of the alias";
            }
        }
        if (item.data) {
            if (datas[item.data] != undefined) {
                datas[item.data]();
            }
        }
        return item;
    }
);


connection.onCodeAction(params => {
    const textDocument = documents.get(params.textDocument.uri);
    if (textDocument == undefined) {
        return undefined
    }
    const [diagnostic] = params.context.diagnostics;
    if (!diagnostic) {
        return []
    }
    const fixes = [];
    if (diagnostic.message.includes("camel case")) {
        const title = "Rename";
        fixes.push(CodeAction.create(title, Command.create(title, 're.name', textDocument.uri, params.range), CodeActionKind.QuickFix))
    }
    return fixes
})

connection.onExecuteCommand(params => {
    switch (params.command) {
        case 're.name': {
            if (params.arguments === undefined) {
                return
            }
            const [document, range] = params.arguments;
            if (!document || !range) {
                return
            }
            const textDocument = documents.get(document);
            if (textDocument === undefined) {
                return;
            }

            const text = textDocument.getText(range);

            const lowered_alias = lowerAlias(text);


            connection.workspace.applyEdit({
                documentChanges: [
                    TextDocumentEdit.create({ uri: textDocument.uri, version: textDocument.version }, [
                        TextEdit.replace(params.arguments[1], lowered_alias)
                    ])
                ]
            });
            break;
        }
        default: {
            break;
        }
    }
})

// connection.onHover(params => {
//     // params.position.line
//     // params.position.character
//     const textDocument = documents.get(params.textDocument.uri);
//     const text = textDocument?.getText();
//     if (text) {
//         const split = text.split("\n");
//         connection.console.log(split[params.position.line]);

//         // const a: Hover = {
//         //     contents: "Teste",
//         //     // range: {
//         //     //     start: {
//         //     //         line: params.position.line,
//         //     //         character: 0
//         //     //     },
//         //     //     end: {
//         //     //         line: params.pos
//         //     //     }
//         //     // }
//         // }
//     }

//     return null;
// })

// Make the text document manager listen on the connection
// for open, change and close text document events
documents.listen(connection);

// Listen on the connection
connection.listen();

server/src/validations.ts
import { setAliasUse, setAliases, AliasInfo, hasDiagnosticRelatedInformationCapability, connection, getDocumentSettings } from "./server";
import {
    Diagnostic,
    DiagnosticSeverity,
    Range,
} from 'vscode-languageserver/node';
import { getOpcode, labels, lowerAlias, parameterCount, removeComments } from "./utils";
import { TextDocument } from 'vscode-languageserver-textdocument';

type Problems = {
    max: number,
    current: number
}

export async function validateTextDocument(textDocument: TextDocument): Promise<void> {
    let settings = await getDocumentSettings(textDocument.uri);
    let problems: Problems = { current: 0, max: settings.maxNumberOfProblems };
    let diagnostics: Diagnostic[] = [];

    // Do the validations
    validateCamelCase(textDocument, problems, diagnostics);
    validateParams(textDocument, problems, diagnostics);
    updateAliases(textDocument, problems, diagnostics);
    validateCommands(textDocument, problems, diagnostics);
    invalidKeywords(textDocument, problems, diagnostics);

    // Send the diagnotics;
    connection.sendDiagnostics({ uri: textDocument.uri, diagnostics });
}

async function validateCommands(textDocument: TextDocument, problems: Problems, diagnostics: Diagnostic[]) {
    const commandPattern = /^ *?robson *.*/gm;
    let match: RegExpExecArray | null;
    const text = textDocument.getText();

    while ((match = commandPattern.exec(text)) && problems.current < problems.max) {
        const stringMatch = match.toString();
        const opcode = getOpcode(stringMatch);
        if (opcode < 0 || opcode > 15) {
            continue
        }
        const start = textDocument.positionAt(match.index);
        const paramsCount = (parameterCount as any)[opcode];
        let given = 0;
        for (let i = 0; i < paramsCount; i++) {
            const range: Range = {
                start: {
                    character: 0,
                    line: start.line + i + 1
                },
                end: {
                    character: 0,
                    line: start.line + i + 2
                }
            }
            const param = removeComments(textDocument.getText(range).trim());
            if (param.length === 0) {
                let diagnostic: Diagnostic = {
                    message: `Missing parameters for ${(labels as any)[opcode]}, needs ${paramsCount} given ${given}`,
                    range,
                    severity: DiagnosticSeverity.Error
                }
                diagnostics.push(diagnostic);
                problems.current++;
                break;
            } else {

            }
            given++
        }
    }
}
async function validateCamelCase(textDocument: TextDocument, problems: Problems, diagnostics: Diagnostic[]) {
    let camelCasePattern = /:[A-Za-z]([A-Z0-9]*[a-z][a-z0-9]*[A-Z]|[a-z0-9]*[A-Z][A-Z0-9]*[a-z])[A-Za-z0-9]*|[A-Za-z]([A-Z0-9]*[a-z][a-z0-9]*[A-Z]|[a-z0-9]*[A-Z][A-Z0-9]*[a-z])[A-Za-z0-9]*:/g;
    let camelCaseMatch: RegExpExecArray | null;
    const text = textDocument.getText();
    while ((camelCaseMatch = camelCasePattern.exec(text)) && problems.current < problems.max) {
        problems.current++;
        let diagnostic: Diagnostic = {
            severity: DiagnosticSeverity.Warning,
            range: rangeFromMatch(camelCaseMatch, textDocument),
            message: `${camelCaseMatch[0]} Has camel case`,
            source: 'ex',
        };

        const lowercase = lowerAlias(camelCaseMatch[0]);

        if (hasDiagnosticRelatedInformationCapability) {
            diagnostic.relatedInformation = [
                {
                    location: {
                        uri: textDocument.uri,
                        range: Object.assign({}, diagnostic.range)
                    },
                    message: `Write it as ${lowercase}`
                },
            ];
        }
        diagnostics.push(diagnostic);
    }
}

async function invalidKeywords(textDocument: TextDocument, problems: Problems, diagnostics: Diagnostic[]) {
    const pattern = /^(?! *(robson|comeu|fudeu|penetrou|chupou|lambeu|;|\w+:|\n)).*/gm;
    let match: RegExpExecArray | null;
    const text = textDocument.getText();
    while ((match = pattern.exec(text)) && problems.current < problems.max) {
        if (match[0].trim().length == 0) {
            break
        }
        let diagnostic: Diagnostic = {
            message: `Unknown keyword ${match.toString()}`,
            range: rangeFromMatch(match, textDocument),
            severity: DiagnosticSeverity.Error
        }
        diagnostics.push(diagnostic);
        problems.current++;
    }
}
function validateParams(textDocument: TextDocument, problems: Problems, diagnostics: Diagnostic[]) {
    const invalidLambeuPattern = /\b *lambeu [^:].*\b/g;
    let invalidLambeuMatch: RegExpExecArray | null;
    const text = textDocument.getText();
    while ((invalidLambeuMatch = invalidLambeuPattern.exec(text)) && problems.current < problems.max) {
        let diagnostic: Diagnostic = {
            severity: DiagnosticSeverity.Error,
            message: "Invalid alias for lambeu",
            range: rangeFromMatch(invalidLambeuMatch, textDocument)
        }
        diagnostics.push(diagnostic);
        problems.current++;
    }
    const invalidComeuPattern = / *comeu (?!([0-9]|f[0-9]|i-?[0-9])).*/g;
    let invalidComeuMatch: RegExpExecArray | null;
    while ((invalidComeuMatch = invalidComeuPattern.exec(text)) && problems.current < problems.max) {
        let diagnostic: Diagnostic = {
            severity: DiagnosticSeverity.Error,
            message: "Invalid number for comeu",
            range: rangeFromMatch(invalidComeuMatch, textDocument)
        }
        problems.current++;
        diagnostics.push(diagnostic);
    }
    const invalidU32Pattern = / *(fudeu|penetrou|chupou) (?![0-9]).*/g
    let invalidU32Match: RegExpExecArray | null;

    while ((invalidU32Match = invalidU32Pattern.exec(text)) && problems.current < problems.max) {
        let name = "";
        const stringMatch = invalidU32Match.toString();

        if (stringMatch.includes("fudeu")) {
            name = "fudeu";
        } else if (stringMatch.includes("penetrou")) {
            name = "penetrou";
        } else {
            name = "chupou"
        }

        let diagnostic: Diagnostic = {
            message: `Invalid number for ${name}`,
            range: rangeFromMatch(invalidU32Match, textDocument),
            severity: DiagnosticSeverity.Error
        }
        problems.current++;
        diagnostics.push(diagnostic);
    }

}


async function updateAliases(textDocument: TextDocument, problems: Problems, diagnostics: Diagnostic[]) {
    const aliasDeclarationPattern = /\w+\b:/g;
    const values: AliasInfo[] = [];
    let aliasDeclarationMatch: RegExpExecArray | null;
    const text = textDocument.getText();
    while ((aliasDeclarationMatch = aliasDeclarationPattern.exec(text)) && problems.current < problems.max) {
        const range = rangeFromMatch(aliasDeclarationMatch, textDocument);
        const stringMatch = aliasDeclarationMatch?.toString().replace(":", "");

        const line = removeComments(textDocument.getText({
            start: {
                character: 0,
                line: range.start.line
            },
            end: range.end
        }));
        if (!line.includes(aliasDeclarationMatch[0])) {
            continue
        }

        const alias = values.find(a => a.value === stringMatch);
        if (alias) {

            let diagnostic: Diagnostic = {
                message: `Duplicated alias ${aliasDeclarationMatch.toString()}`,
                range,
                severity: DiagnosticSeverity.Error
            }
            if (hasDiagnosticRelatedInformationCapability) {
                diagnostic.relatedInformation = [
                    {
                        location: {
                            uri: textDocument.uri,
                            range: Object.assign({}, alias.range)
                        },
                        message: "First declared here"
                    },
                ];
            }
            diagnostics.push(diagnostic);
            problems.current++;
        } else {
            values.push({ range, value: stringMatch });
        }
    }
    setAliases(values);

    const aliasUsagePattern = /\w*:\b\w+/g
    let aliasUsageMatch: RegExpExecArray | null;
    let aliasUsages: AliasInfo[] = [];
    while (aliasUsageMatch = aliasUsagePattern.exec(text)) {
        let stringMatch = aliasUsageMatch?.toString().replace(":", "");
        const declaration = values.find(a => a.value == stringMatch);
        const range: Range = rangeFromMatch(aliasUsageMatch, textDocument);
        const line = removeComments(textDocument.getText({
            start: {
                character: 0,
                line: range.start.line
            },
            end: range.end
        }));
        if (!line.includes(aliasUsageMatch[0])) {
            continue
        }
        if (!declaration) {
            let diagnostic: Diagnostic = {
                message: "Using alias before declaration",
                range,
                severity: DiagnosticSeverity.Error
            }
            diagnostics.push(diagnostic);
            problems.current++;
        }
        aliasUsages.push({ value: stringMatch, range });
    }

    for (const aliasDeclaration of values) {
        const finded = aliasUsages.find(a => a.value == aliasDeclaration.value);
        if (!finded) {
            let diagnostic: Diagnostic = {
                message: "This alias is never used",
                range: aliasDeclaration.range,
                severity: DiagnosticSeverity.Warning
            }
            diagnostics.push(diagnostic);
            problems.current++;
        }
    }

    setAliasUse(aliasUsages)
}
function rangeFromMatch(match: RegExpExecArray, textDocument: TextDocument): Range {
    return {
        start: textDocument.positionAt(match.index),
        end: textDocument.positionAt(match.index + match[0].length),
    }
}

src/extension.ts
import * as path from "path";
import { LanguageClient, LanguageClientOptions, ServerOptions, TransportKind } from "vscode-languageclient/node";
import { ExtensionContext, workspace } from "vscode";
import * as vscode from "vscode";
let client: LanguageClient;

function getFileName(path: string): string {
  const splited = path.split("/");
  const windowsSplited = path.split("\\");

  if (splited.length > windowsSplited.length) {
    if (splited.length === 0) {
      return path
    }
    return splited[splited.length - 1]
  } else {
    return windowsSplited[windowsSplited.length - 1]
  }
}

function removeComments(string: string): string {
  const removed_comments = string.split(";");
  if (removed_comments.length > 0) {
    string = removed_comments[0];
  }
  return string.trim()
}


function getOpcode(string: string): number {
  let opcode = 0;
  const removed_comments = string.split(";");
  if (removed_comments.length > 0) {
    string = removed_comments[0];
  }
  const splited = string.trim().split(' ');
  for (const keyword of splited) {
    if (keyword.trim() == "robson") {
      opcode++;
    } else {
      opcode = -1;
      break;
    }
  }
  return opcode;
}

function multi_function_opcode(object: any, max: number, lines: string[], i: number): string {
  let label = "";
  let next = lines[i + 1];
  next = removeComments(next);
  next = next.trim();
  if (next.length != 0) {
    const splited = next.split(" ");
    if (splited.length > 0) {
      const value = Number(splited[1]);
      if (value < max) {
        label = object[value];
      }
    }
  }
  return label
}

const labels = {
  1: "operation",
  2: "if lower",
  3: "push",
  4: "if equal",
  5: "vstack",
  6: "input",
  7: "print",
  8: "pnumber",
  9: "jump",
  10: "set",
  11: "pop",
  12: "load",
  13: "time",
  14: "flush",
  15: "terminal"
}
const operations = {
  0: "add",
  1: "sub",
  2: "mul",
  3: "div"
}
const time = {
  0: "count",
  1: "interval",
  2: "cmp_interval"
}
const terminal = {
  0: "raw",
  1: "clear",
  2: "poll",
  3: "togg_cursor",
  4: "mv_cursor",
  5: "color"
}

export function activate(context: ExtensionContext) {
  console.log("teste");

  // The server is implemented in node
  let serverModule = context.asAbsolutePath(path.join('server', 'out', 'server.js'));
  // The debug options for the server
  // --inspect=6009: runs the server in Node's Inspector mode so VS Code can attach to the server for debugging
  let debugOptions = { execArgv: ['--nolazy', '--inspect=6009'] };

  // If the extension is launched in debug mode then the debug server options are used
  // Otherwise the run options are used
  let serverOptions: ServerOptions = {
    run: { module: serverModule, transport: TransportKind.ipc },
    debug: {
      module: serverModule,
      transport: TransportKind.ipc,
      options: debugOptions
    }
  };

  // Options to control the language client
  let clientOptions: LanguageClientOptions = {
    // Register the server for plain text documents
    documentSelector: [{ scheme: 'file', language: 'robson' }],
    synchronize: {
      // Notify the server about file changes to '.clientrc files contained in the workspace
      fileEvents: workspace.createFileSystemWatcher('**/.clientrc')
    }
  };

  // Create the language client and start the client.
  client = new LanguageClient(
    'RobsonAnalyzer',
    'Robson Analyzer Server',
    serverOptions,
    clientOptions
  );

  // Start the client
  client.start();


  const provider: vscode.InlayHintsProvider<vscode.InlayHint> = {

    provideInlayHints(document, position, context) {

      const text = document.getText();
      const hints: vscode.InlayHint[] = [];

      const lines = text.split("\n");

      console.log(lines);
      console.log(text);

      let index = 0;
      for (let i = 0; i < lines.length; i++) {
        const line = lines[i];

        if (line.includes("robson")) {
          const opcode = getOpcode(line);
          if (opcode == -1) {
            index += line.length + 1;
            continue
          }
          let label = (opcode < 1 || opcode > 15) ? "unknown" : labels[opcode];
          let new_label = "";
          if (opcode == 1) {
            new_label = multi_function_opcode(operations, 4, lines, i);
          } else if (opcode == 13) {
            new_label = multi_function_opcode(time, 3, lines, i);
          } else if (opcode == 15) {
            new_label = multi_function_opcode(terminal, 5, lines, i);
          }
          if (new_label != "") {
            label = new_label;
          }
          let before_spaces = 0;
          for (const char of line) {
            if (char === " ") {
              before_spaces++
            } else {
              break
            }
          }

          const position = document.positionAt(index + before_spaces);
          const hint: vscode.InlayHint = {
            label,
            position,
            kind: vscode.InlayHintKind.Type,
            paddingRight: true
          }
          hints.push(hint);
        }
        index += line.length + 1;
      }
      return hints as vscode.ProviderResult<vscode.InlayHint[]>;
    }
  }

  vscode.languages.registerInlayHintsProvider({ pattern: "**", language: "robson" }, provider);

  vscode.languages.registerHoverProvider("robson", {
    provideHover(textDocument, position, c) {
      const wordRange = textDocument.getWordRangeAtPosition(position);
      const word = textDocument.getText(wordRange);
      let line = textDocument.getText(new vscode.Range(new vscode.Position(position.line, 0), new vscode.Position(position.line + 1, 0)))
      // if (word === "robson") {
      //   return {
      //     contents: ["robsion"]
      //   };
      // }
      line = removeComments(line);

      if (line.length < position.character) {
        return null
      }

      if (word === "comeu") {
        return {
          contents: [`# Comeu
          Get the raw value you input, the simpler you can get
            
            robson robson robson
            comeu 10
          
          This will simply push the number 10

          For floats or signed integers use the following respectively

          robson robson robson
          comeu f10 ; for floats

          robson robson robson
          comeu i10 ; for signed integers 

          `]
        }
      }

      if (word === "chupou") {
        return {
          contents: [`# Chupou
          Get the value of the top of the stack
            
            robson robson robson
            chupou 0
          
          chupou will not work if the value is other than 0
          `]
        }
      }

      if (word === "fudeu") {
        return {
          contents: [`# Fudeu
          Get the value in the given memory address
            
            robson robson robson
            fudeu 51
          
          fudeu only accepts usigned integers
          `]
        }
      }
      if (word === "penetrou") {
        return {
          contents: [`# Penetrou
          Follows the memory address 2 times
            
            robson robson robson
            penetrou 2
          
          In this exmaple it will push the value in the address that the address 2 points

          penetrou only accepts unisgned integers
          `]
        }
      }

      if (word === "lambeu") {
        return {
          contents: [`# Lambeu
          Get the value of the number of the next command in an alias
            
            alias: 

            robson robson robson
            lambeu :alias
          
          This example will push 0, because the alias is behind the first command
          `]
        }
      }


      return null
    }
  })


  let type = "robson";
  vscode.tasks.registerTaskProvider(type, {
    provideTasks() {
      const path = vscode.window.activeTextEditor.document.uri.fsPath;
      vscode.window.activeTextEditor.document.save();
      let execution = new vscode.ShellExecution(`robson ${path} compile`);
      let task = new vscode.Task({ type }, vscode.TaskScope.Workspace,
        `Compile ${getFileName(path)}`, "Robson", execution);
      task.group = vscode.TaskGroup.Build;
      return [
        task
      ];
    },
    resolveTask(task) {
      return task;
    }
  });
}

export function deactivate(): Thenable<void> | undefined {
  if (!client) {
    return undefined
  }
  return client.stop()
}


